src/interpreter.ts

```diff
@@ -1438,6 +1442,53 @@ export class Interpreter {
         this.envStack.setNewBinding(idText(ast.name), val);
     }
 
+    public interpretDestructStatement(ast: AstStatementDestruct) {
+        for (const item of ast.identifiers) {
+            if (hasStaticConstant(this.context, idText(item.name))) {
+                // Attempt of shadowing a constant in a destructuring declaration
+                throwInternalCompilerError(
+                    `declaration of ${idText(item.name)} shadows a constant with the same name`,
+                    ast.loc,
+                );
+            }
+        }
+        const val = this.interpretExpression(ast.expression);
+        if (
+            val === null ||
+            typeof val !== "object" ||
+            !("$tactStruct" in val)
+        ) {
+            throwErrorConstEval(
+                `destructuring assignment expected a struct, but got ${showValue(
+                    val,
+                )}`,
+                ast.expression.loc,
+            );
+        }
+        if (ast.identifiers.length !== Object.keys(val).length - 1) {
+            throwErrorConstEval(
+                `destructuring assignment expected ${ast.identifiers.length} fields, but got ${
+                    Object.keys(val).length - 1
+                }`,
+                ast.loc,
+            );
+        }
+
+        for (const item of ast.identifiers) {
+            const v = val[idText(item.from)];
+            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+            if (v === undefined) {
+                throwErrorConstEval(
+                    `destructuring assignment expected field ${idTextErr(
+                        item.from,
+                    )}`,
+                    ast.loc,
+                );
+            }
+            this.envStack.setNewBinding(idText(item.name), v);
+        }
+    }
+
```

Comment:
The interpreter code looks good. For positive tests, I suggest adding tests of the following form:
```
struct S {
  a: Int;
  b: Int;
  c: Int; 
}

fun testFun1(): Int {
    let s = S { a: 1, b: 2, c: 3 };
    let S { c: e, b, a } = s;
    return a + b + e;
}

fun testFun2(): S {
    let s = S { a: 1, b: 2, c: 3 };
    let S { a, c: e, b: b } = s;
    return S {a: e, b: b, c: a};
}

contract TestContract {
    get fun test1(): Int {
        return testFun1();
    }
    
    get fun test2(): S {
        return testFun2();
    }
}
```
In other words, just repeat the tests you already have, but this time call the functions in a getter, since this will force the execution of the interpreter in the function calls inside the getters.

Negative tests for the interpreter are more difficult to write because the errors are caught by the typechecker before they arrive to the interpreter. I have plans for testing the interpreter in isolation, but for the moment, just make sure that in the negative tests the typechecker gets errors for:
- Constant shadowing:
   ```
   const c1: Int = 5;
   .........
   let S { a: c1, b, c } = s;   // c1 is shadowing constant c1
   ```
- Destructuring of non-struct value:
   ```
   let S { a, b, c } = i;    // where i is an integer
   ```
- Missing and extra fields on both sides of the equality (which I think this is already tested):
   ```
   struct A {
     f1: Int;
     f2: Int;
   }

   struct B {
      f1: Int;
      f2: Int;  
   }
   ....... Take each line as an independent test (also, repeat the tests using non-punned versions):
   let A {f1} = s  // s has type A
   let A {f1, f2} = A {f1: 2};
   let A {f1, f2} = s  // s has type B, even though they have the same fields, they differ in types
   let A {f1, f2, f3} = s  // s has type A
   let A {f1, f2} = A {f1: 2, f3: 6};
   ```
- Plus all tests suggested by Anton.

The interpreter makes the assumption that the code it receives has already passed thorough the typechecker. Otherwise, we would repeat a lot of code.

Reply:
I would also initialize some constants using functions with `let`-destructuring